From 2875428f3fb16c2e21dfb9f72de84bf69eaf6702 Mon Sep 17 00:00:00 2001
From: wuzhouhui <wuzhouhui250@gmail.com>
Date: Sat, 2 Mar 2019 12:07:33 +0800
Subject: [PATCH 91/95] svn/ci: support client side hooks, only pre-commit for
 now

The hook scripts is located in .svn/hooks, and the naming convention of
hook scripts is the same as server side.
---
 README                      |  2 ++
 subversion/svn/commit-cmd.c | 56 +++++++++++++++++++++++++++++++++++++++++++++
 2 files changed, 58 insertions(+)

diff --git a/README b/README
index 1465fd8..bfea9ce 100644
--- a/README
+++ b/README
@@ -8,3 +8,5 @@ Main patches:
       5. Add option -v for subcommand commit, for print unified diff in log
          message template
       6. Add option --stat for subcommand diff and log
+      7. Support client side pre-commit hook, hook scripts locate in
+         .svn/hooks
diff --git a/subversion/svn/commit-cmd.c b/subversion/svn/commit-cmd.c
index 9701697..9f09d62 100644
--- a/subversion/svn/commit-cmd.c
+++ b/subversion/svn/commit-cmd.c
@@ -89,6 +89,59 @@ copy_warning_notify_func(void *baton,
     }
 }
 
+static svn_error_t *
+pre_commit(svn_client_ctx_t *ctx, const char *target, apr_pool_t *pool)
+{
+  const char *wcroot_abspath, *abspath;
+  svn_node_kind_t kind;
+  svn_boolean_t executable = FALSE;
+  int sys_err;
+
+  SVN_ERR(svn_dirent_get_absolute(&abspath, target, pool));
+
+  SVN_ERR(svn_client_get_wc_root(&wcroot_abspath,
+                                 abspath,
+                                 ctx,
+                                 pool,
+                                 pool));
+
+  abspath = svn_dirent_join_many(pool,
+                                 wcroot_abspath,
+                                 svn_wc_get_adm_dir(pool),
+                                 "hooks/pre-commit",
+                                 SVN_VA_NULL);
+
+  /* Now, the abspath points to pre-commit hook script. If script
+   * exist and executable, we will run it.
+   */
+
+  SVN_ERR(svn_io_check_path(abspath, &kind, pool));
+  if (kind == svn_node_none)
+    {
+      return SVN_NO_ERROR;
+    }
+  if (kind != svn_node_file)
+    {
+      return svn_error_create(0, NULL, _("pre-commit hook is not a file"));
+    }
+
+  SVN_ERR(svn_io_is_file_executable(&executable, abspath, pool));
+  if (!executable)
+    {
+      return svn_error_create(0, NULL,
+                              _("pre-commit hook is not executable"));
+    }
+
+  /* execute pre-commit */
+  sys_err = system(abspath);
+  if (sys_err != 0)
+    {
+      return svn_error_create(SVN_ERR_EXTERNAL_PROGRAM, NULL,
+                              _("pre-commit hook exits with failure"));
+    }
+
+  return SVN_NO_ERROR;
+}
 
 
 
@@ -118,6 +171,9 @@ svn_cl__commit(apr_getopt_t *os,
   /* Add "." if user passed 0 arguments. */
   svn_opt_push_implicit_dot_target(targets, pool);
 
+  /* Execute pre-commit hook if possible */
+  SVN_ERR(pre_commit(ctx, APR_ARRAY_IDX(targets, 0, const char *), pool));
+
   SVN_ERR(svn_cl__eat_peg_revisions(&targets, targets, pool));
 
   /* Condense the targets (like commit does)... */
-- 
1.8.3.1

